File: java/03_chapter/BST.java
/**
 * 基于二叉查找树的符号表
 */
public class BST<Key extends Comparable<Key>, Value> {
    private Node node;
    private class Node {
        private Key key;
        private Value val;
        private Node left, right;
        private int N;
        public Node(Key key, Value val, int N){
            this.key = key;
            this.val = val;
            this.N = N;
        }

        public int size(){
            return size(root);
        }

        public int size(Node x){
            if ( x == null) return 0;
            else return x.N;
        }

        public Value get(Key key){
            return get(root, key);
        }

        public Value get(Node x, Key key){
            if (x == null) {
                return null;
            }
            int cmp = key.compareTo(x.key){
                if (cmp < 0) {
                    return get(x.left, key);
                } else if (cmp > 0){
                    return get(x.right, key);
                } else {
                    return x.val;
                }
            }
        }

        public void put(Key key, Value val){
            root = put(root, key, val);
        }

        public Node put(Node x, Key key, Value val){
            if (x == null) return new Node(key, val, 1);
            int cmp = key.compareTo(x.key)
            if (cmp < 0){
                x.left = put(x.left, key, val);
            } else if (cmp > 0){
                x.right = put(x.right, key, val);
            } else {
                x.val = val;
            }
            x.N = size(x.left) + size(x.right) + 1;
        }
    }
}

File: javascript/03_chapter/BST.js
/**
 * 基于二叉查找树的符号表
 */

class Node {
    constructor(key, val, N){
        this.key = key;
        this.val = val;
        this.N = N;
    }
}

class BST {
    constructor(){
        this.root = null;
        this.left = null;
        this.right = null;
    }

    size(node){
        if (node){
            return node.N;
        } else {
            return 0;
        }
    }

    get(key){
        this._get(this.root, key)
    }

    _get(node, key){
        if (node === null){
            return null;
        }
        if (node.key < key){
            return this._get(node.right, key)
        } else if (node.key > key){
            return this._get(node.left, key);
        } else {
            return node.val;
        }
    }

    put(key, val){
       this.root = this._put(this.root, key, val);
    }

    _put(node, key, val){
        if (!node){
            return new Node(key, val, 1);
        }
        if (node.key < key){
            node.right = this._put(node.right, key, val)
        } else if (node.key > key){
            node.left = this._put(node.left, key, val)
        } else {
            node.val = val;
        }
        node.N = this.size(node.left) + this.size(node.right) + 1;
        return node;
    }

    print(){
        this._print(this.root);
    }

    _print(node){
        if (node){
            this._print(node.left);
            console.log(node.key);
            this._print(node.right);
        }
    }
}

let st = new BST();
st.put('L', 11);
st.put('P', 10);
st.put('M', 9);
st.put('X', 7);
st.put('H', 5);
st.put('C', 4);
st.put('R', 3);
st.put('A', 8);
st.put('E', 12);
st.put('S', 0);

st.print();


// console.log(st.root.val, st.root.left.val, st.root.right.val);
// console.log(st.get('A'))
// console.log(st.get('C'))
// console.log(st.get('M'))
// console.log(st.get('S'))
// console.log(st.get('X'))